File: pbhhg_py/parse.py
"""Parser for esoteric language Unsuspected Hangeul."""
import re
import unicodedata

from pbhhg_py import abstract_syntax as AS

# TABLES
# note all ㅇ and ㅎ has a preceding space in following tables
U1100 = (
    "ㄱ|ㄱ|ㄴ|ㄷ|ㄷ|ㄹ|ㅁ|ㅂ|ㅂ|ㅅ|ㅅ| ㅇ|ㅈ|ㅈ|ㅈ|ㄱ|ㄷ|ㅂ| ㅎ|ㄴㄱ|ㄴ|ㄴㄷ|ㄴㅂ|ㄷㄱ|ㄹㄴ|ㄹ|"
    "ㄹ ㅎ|ㄹ|ㅁㅂ|ㅁ|ㅂㄱ|ㅂㄴ|ㅂㄷ|ㅂㅅ|ㅂㅅㄱ|ㅂㅅㄷ|ㅂㅅㅂ|ㅂㅅ|ㅂㅅㅈ|ㅂㅈ|ㅂㅈ|ㅂㄷ|ㅂㅂ|ㅂ|"
    "ㅂ|ㅅㄱ|ㅅㄴ|ㅅㄷ|ㅅㄹ|ㅅㅁ|ㅅㅂ|ㅅㅂㄱ|ㅅㅅ|ㅅ ㅇ|ㅅㅈ|ㅅㅈ|ㅅㄱ|ㅅㄷ|ㅅㅂ|ㅅ ㅎ|ㅅ|ㅅ|ㅅ|ㅅ|"
    "ㅅ| ㅇㄱ| ㅇㄷ| ㅇㅁ| ㅇㅂ| ㅇㅅ| ㅇㅅ| ㅇ| ㅇㅈ| ㅇㅈ| ㅇㄷ| ㅇㅂ| ㅇ|ㅈ ㅇ|ㅈ|ㅈ|ㅈ|ㅈ|"
    "ㅈㄱ|ㅈ ㅎ|ㅈ|ㅈ|ㅂㅂ|ㅂ| ㅎ| ㅎ|ㄱㄷ|ㄴㅅ|ㄴㅈ|ㄴ ㅎ|ㄷㄹ"
).split("|")
JAMO = (
    "ㄱ|ㄱ|ㄱㅅ|ㄴ|ㄴㅈ|ㄴ ㅎ|ㄷ|ㄷ|ㄹ|ㄹㄱ|ㄹㅁ|ㄹㅂ|ㄹㅅ|ㄹㄷ|ㄹㅂ|ㄹ ㅎ|ㅁ|ㅂ|ㅂ|ㅂㅅ|ㅅ|ㅅ|"
    " ㅇ|ㅈ|ㅈ|ㅈ|ㄱ|ㄷ|ㅂ| ㅎ"
).split("|")
U3165 = (
    "ㄴ|ㄴㄷ|ㄴㅅ|ㄴㅅ|ㄹㄱㅅ|ㄹㄷ|ㄹㅂㅅ|ㅁㅅ|ㅁ|ㅂㄱ|ㅂㄷ|ㅂㅅㄱ|ㅂㅅㄷ|ㅂㅈ|ㅂㄷ|ㅂ|ㅂ|ㅅㄱ|ㅅㄴ|"
    " ㅇ| ㅇ| ㅇㅅ| ㅇㅅ|ㅂ| ㅎ| ㅎ"
).split("|")
UA960 = (
    "ㄷㅁ|ㄷㅂ|ㄷㅅ|ㄷㅈ|ㄹㄱ|ㄹㄱ|ㄹㄷ|ㄹㄷ|ㄹㅁ|ㄹㅂ|ㄹㅂ|ㄹㅂ|ㄹㅅ|ㄹㅈ|ㄹㄱ|ㅁㄱ|ㅁㄷ|ㅁㅅ|"
    "ㅂㅅㄷ|ㅂㄱ|ㅂ ㅎ|ㅅㅂ| ㅇㄹ| ㅇ ㅎ|ㅈ ㅎ|ㄷ|ㅂ ㅎ| ㅎㅅ| ㅎ"
).split("|")


def normalize_char(c: str):
    """Normalizes each character into standard form"""
    if len(c) != 1:
        raise ValueError(
            "Length of string should be 1, not {}: {}".format(len(c), c)
        )

    def _get(arr: list[str], offset: str):
        idx = ord(c) - ord(offset)
        if 0 <= idx < len(arr):
            return arr[idx]
        return ""

    if "\u1100" <= c <= "\u11FF":  # 첫가끝
        return _get(U1100, "\u1100")
    if "\u302E" <= c <= "\u302F":  # 방점
        return ""
    if "\u3131" <= c <= "\u3164":  # 호환용 (현대 한글)
        return _get(JAMO, "\u3131")
    if "\u3165" <= c <= "\u318E":  # 호환용 (옛한글)
        return _get(U3165, "\u3165")
    if "\uA960" <= c <= "\uA97C":  # 확장A
        return _get(UA960, "\uA960")
    if "\uD7B0" <= c <= "\uD7C6":  # 확장B
        return ""
    if "\uD7CB" <= c <= "\uD7FB":
        return ""
    if "\uFFA1" <= c <= "\uFFBE":  # 반각
        return _get(JAMO, "\uFFA1")
    if c in "ￂￃￄￅￆￇￊￋￌￍￎￏￒￓￔￕￖￗￚￛￜ":
        return ""
    return " "


def normalize(character: str) -> list[str]:
    characters = unicodedata.normalize("NFD", character)
    normalized = "".join(normalize_char(c) for c in characters)
    return re.split(r"( )", normalized)


def _merge_metadata(a: AS.Metadata, b: AS.Metadata) -> AS.Metadata:
    assert a.filename == b.filename
    assert a.line_no == b.line_no
    return AS.Metadata(a.filename, a.line_no, a.start_col, b.end_col, a.line)


def tokenize(filename: str, sentence: str) -> list[tuple[str, AS.Metadata]]:
    lines = sentence.split("\n")
    characters = [
        (d, AS.Metadata(filename, i, j, j + 1, line))
        for i, line in enumerate(lines)
        for j, c in enumerate(line + "\n")
        for d in normalize(c)
        if d
    ]
    if not characters:
        return []

    tokens: list[tuple[str, AS.Metadata]] = [characters[0]]
    for cur, metadata in characters[1:]:
        prev, prev_metadata = tokens[-1]
        if prev == " " and cur == " ":
            continue
        elif prev != " " and cur != " ":
            tokens[-1] = (prev + cur, _merge_metadata(prev_metadata, metadata))
        else:
            tokens.append((cur, metadata))
    return [token for token in tokens if token[0] != " "]


def parse_number(s: str):
    """Parses jamo-encoded variable length integer into python int"""
    tbl = "ㄱㄴㄷㄹㅁㅂㅅㅈ"
    varlen = [tbl.index(c) for c in s]  # LSB first
    octal = "".join([str(d) for d in reversed(varlen)])  # MSB first
    num = int(octal, base=8)
    if len(s) % 2 == 0:
        num = -num
    return num


def encode_number(number: int):
    """Jamo-encodes an python int."""
    is_negative = number < 0
    if is_negative:
        number = -number
    # nonnegative integer
    octal: list[int] = [int(d) for d in f"{number:o}"]
    tbl = "ㄱㄴㄷㄹㅁㅂㅅㅈ"
    encoded = [tbl[s] for s in reversed(octal)]
    if (len(encoded) % 2 == 0) != is_negative:
        encoded.append("ㄱ")
    return "".join(encoded)


def parse_token(
    token: tuple[str, AS.Metadata], stack: list[AS.AST]
) -> list[AS.AST]:
    """Parses concrete syntax to abstract syntax
    Args:
        word: token to parse
        stack: list of parsed legal arguments so far

    Returns:
        new stack with newly parsed argument appended
    """
    word, metadata = token
    if "ㅎ" in word:
        _, arity = word.split("ㅎ")

        if arity:  # FunCall
            arity = parse_number(arity)
            if arity < 0:
                raise SyntaxError(f"함수 호출 시 {arity}개의 인수를 요구했습니다.")

            fun = stack.pop()
            argv = stack[-arity:] if arity else []
            rest = stack[:-arity] if arity else stack
            if len(argv) < arity:
                raise SyntaxError(
                    f"함수 호출 시 {arity}개의 인수를 요구했으나 "
                    f"표현식이 {len(argv)}개밖에 없습니다."
                )

            return rest + [AS.FunCall(fun, tuple(argv), metadata)]

        else:  # FunDef
            body = stack.pop()
            return stack + [AS.FunDef(body, metadata)]

    elif "ㅇ" in word:
        _, trailer = word.split("ㅇ")

        if trailer:  # ArgRef
            relF = parse_number(trailer)
            relA = stack.pop()
            return stack + [AS.ArgRef(relA, relF, metadata)]

        else:  # FunRef
            relF = stack.pop()
            if not isinstance(relF, AS.Literal):
                raise SyntaxError(f"함수 참조 시에는 정수 리터럴만 허용되는데 {relF}를 받았습니다.")
            relF = relF.value
            return stack + [AS.FunRef(relF, metadata)]

    else:
        return stack + [AS.Literal(parse_number(word), metadata)]


def parse(filename: str, sentence: str):
    """Parses program into abstract syntax"""
    tokens = tokenize(filename, sentence)
    stack = []
    for token in tokens:
        stack = parse_token(token, stack)
    return stack

File: pbhhg_js/parse.ts
/* Parser for unsuspected hangeul. */
import * as AS from './abstractSyntax.js'

/** Normalizer **/

// TABLES
// note all ㅇ and ㅎ has a preceding space in following tables
const U1100 = (
  'ㄱ|ㄱ|ㄴ|ㄷ|ㄷ|ㄹ|ㅁ|ㅂ|ㅂ|ㅅ|ㅅ| ㅇ|ㅈ|ㅈ|ㅈ|ㄱ|ㄷ|ㅂ| ㅎ|ㄴㄱ|ㄴ|ㄴㄷ|ㄴㅂ|ㄷㄱ|ㄹㄴ|ㄹ|' +
  'ㄹ ㅎ|ㄹ|ㅁㅂ|ㅁ|ㅂㄱ|ㅂㄴ|ㅂㄷ|ㅂㅅ|ㅂㅅㄱ|ㅂㅅㄷ|ㅂㅅㅂ|ㅂㅅ|ㅂㅅㅈ|ㅂㅈ|ㅂㅈ|ㅂㄷ|ㅂㅂ|ㅂ|' +
  'ㅂ|ㅅㄱ|ㅅㄴ|ㅅㄷ|ㅅㄹ|ㅅㅁ|ㅅㅂ|ㅅㅂㄱ|ㅅㅅ|ㅅ ㅇ|ㅅㅈ|ㅅㅈ|ㅅㄱ|ㅅㄷ|ㅅㅂ|ㅅ ㅎ|ㅅ|ㅅ|ㅅ|ㅅ|' +
  'ㅅ| ㅇㄱ| ㅇㄷ| ㅇㅁ| ㅇㅂ| ㅇㅅ| ㅇㅅ| ㅇ| ㅇㅈ| ㅇㅈ| ㅇㄷ| ㅇㅂ| ㅇ|ㅈ ㅇ|ㅈ|ㅈ|ㅈ|ㅈ|' +
  'ㅈㄱ|ㅈ ㅎ|ㅈ|ㅈ|ㅂㅂ|ㅂ| ㅎ| ㅎ|ㄱㄷ|ㄴㅅ|ㄴㅈ|ㄴ ㅎ|ㄷㄹ'
).split('|')
const JAMO = (
  'ㄱ|ㄱ|ㄱㅅ|ㄴ|ㄴㅈ|ㄴ ㅎ|ㄷ|ㄷ|ㄹ|ㄹㄱ|ㄹㅁ|ㄹㅂ|ㄹㅅ|ㄹㄷ|ㄹㅂ|ㄹ ㅎ|ㅁ|ㅂ|ㅂ|ㅂㅅ|ㅅ|ㅅ|' +
  ' ㅇ|ㅈ|ㅈ|ㅈ|ㄱ|ㄷ|ㅂ| ㅎ'
).split('|')
const U3165 = (
  'ㄴ|ㄴㄷ|ㄴㅅ|ㄴㅅ|ㄹㄱㅅ|ㄹㄷ|ㄹㅂㅅ|ㅁㅅ|ㅁ|ㅂㄱ|ㅂㄷ|ㅂㅅㄱ|ㅂㅅㄷ|ㅂㅈ|ㅂㄷ|ㅂ|ㅂ|ㅅㄱ|ㅅㄴ|' +
  ' ㅇ| ㅇ| ㅇㅅ| ㅇㅅ|ㅂ| ㅎ| ㅎ'
).split('|')
const UA960 = (
  'ㄷㅁ|ㄷㅂ|ㄷㅅ|ㄷㅈ|ㄹㄱ|ㄹㄱ|ㄹㄷ|ㄹㄷ|ㄹㅁ|ㄹㅂ|ㄹㅂ|ㄹㅂ|ㄹㅅ|ㄹㅈ|ㄹㄱ|ㅁㄱ|ㅁㄷ|ㅁㅅ|' +
  'ㅂㅅㄷ|ㅂㄱ|ㅂ ㅎ|ㅅㅂ| ㅇㄹ| ㅇ ㅎ|ㅈ ㅎ|ㄷ|ㅂ ㅎ| ㅎㅅ| ㅎ'
).split('|')

/** Normalizes each character into standard form */
function normalizeChar(c: string) {
  if (c.length !== 1) {
    throw Error(
      '[normalizeChar] Length of string should be 1, not ' + c.length + ': ' + c
    )
  }

  function _get(arr: string[], ref: string, divisor = 1) {
    let idx = c.charCodeAt(0) - ref.charCodeAt(0)
    idx = Math.floor(idx / divisor)
    if (idx >= 0 && idx < arr.length) {
      return arr[idx]
    }
    return ''
  }

  if (c >= '\u1100' && c <= '\u11FF') {
    return _get(U1100, '\u1100')
  } else if (c >= '\u302E' && c <= '\u302F') {
    return ''
  } else if (c >= '\u3131' && c <= '\u3164') {
    return _get(JAMO, '\u3131')
  } else if (c >= '\u3165' && c <= '\u318E') {
    return _get(U3165, '\u3165')
  } else if (c >= '\uA960' && c <= '\uA97C') {
    return _get(UA960, '\uA960')
  } else if (c >= '\uAC00' && c <= '\uD7AF') {
    return _get(U1100, '\uAC00', 588)
  } else if (c >= '\uD7B0' && c <= '\uD7C6') {
    return ''
  } else if (c >= '\uD7CB' && c <= '\uD7FB') {
    return ''
  } else if (c >= '\uFFA0' && c <= '\uFFBE') {
    return _get(JAMO, '\uFFA1')
  } else if ('ￂￃￄￅￆￇￊￋￌￍￎￏￒￓￔￕￖￗￚￛￜ'.indexOf(c) >= 0) {
    return ''
  } else return ' '
}

export function normalize(sentence: string) {
  const normalized = sentence.split('').map(normalizeChar).join('').trim()
  return normalized.split(/( )/)
}

function mergeMetadata(a: AS.Metadata, b: AS.Metadata) {
  return new AS.Metadata(a.filename, a.line_no, a.start_col, b.end_col, a.line)
}

function tokenize(filename: string, sentence: string) {
  if (sentence === '') return []
  const lines = sentence.split('\n')
  const characters = lines.flatMap((line, i) =>
    (line + '\n').split('').flatMap((c, j) =>
      normalize(c)
        .filter((d) => d !== '')
        .map((d): [string, AS.Metadata] => [
          d,
          new AS.Metadata(filename, i, j, j + 1, line),
        ])
    )
  )
  let tokens = [characters[0]]
  for (const [cur, metadata] of characters.slice(1)) {
    const top = tokens.pop()
    if (top == null) throw Error('Internal::tokenize::EMPTY_STACK')
    const [prev, prevMetadata] = top
    if (prev === ' ' && cur === ' ') {
      tokens.push([prev, prevMetadata])
    } else if (prev !== ' ' && cur === ' ') {
      tokens.push([prev + cur, mergeMetadata(prevMetadata, metadata)])
    } else {
      tokens.push([cur, metadata])
    }
  }
  return tokens.filter(([c, metadata]) => c !== ' ')
}

/* Parses jamo-encoded variable length integer into JS Number */
export function parseNumber(s: string): bigint {
  const tbl = 'ㄱㄴㄷㄹㅁㅂㅅㅈ'
  const varlen = s.split('').map(function (c) {
    return tbl.indexOf(c)
  })
  if (varlen.indexOf(-1) !== -1) {
    throw SyntaxError('Argument ' + s + ' has an unsupported character')
  }
  let num = BigInt('0o' + varlen.reverse().join(''))
  if (s.length % 2 === 0) {
    num = -num
  }
  return num
}

/* Encodes JS integer to jamo-encoded variable length integer. */
export function encodeNumber(number: number | bigint) {
  const isNegative = number < 0
  number = number >= 0 ? number : -number
  const oct = number.toString(8).split('').reverse()
  const tbl = 'ㄱㄴㄷㄹㅁㅂㅅㅈ'
  let encoded = oct.map((s) => tbl[+s]).join('')
  if ((encoded.length % 2 === 0) !== isNegative) {
    encoded += 'ㄱ'
  }
  return encoded
}

/** Parses concrete syntax to abstract syntax
 * @param token token to parse
 * @param stack list of parsed legal arguments so far that we will modify
 */
function parseToken(token: [string, AS.Metadata], stack: AS.AST[]) {
  const [word, metadata] = token
  if (word.indexOf('ㅎ') !== -1) {
    const arity = word.slice(1)

    if (arity) {
      // AS.FunCall
      const _arity = Number(parseNumber(arity))
      if (_arity < 0) {
        throw SyntaxError(`함수 호출 시 ${_arity}개의 인수를 요구했습니다.`)
      }

      const fun = stack.pop()
      if (fun == null) throw Error('Internal::parseToken::EMPTY_STACK')
      if (_arity === 0) {
        stack.push(new AS.FunCall(metadata, fun, []))
      } else {
        const argv = stack.splice(-_arity, _arity)
        if (argv.length < _arity) {
          throw SyntaxError(
            `함수 호출 시 ${_arity}개의 인수를 요구했으나 표현식이 ${argv.length}개밖에 없습니다.`
          )
        }
        stack.push(new AS.FunCall(metadata, fun, argv))
      }
    } else {
      // AS.FunDef
      const body = stack.pop()
      if (body == null) throw Error('Internal::parseToken::EMPTY_STACK')
      stack.push(new AS.FunDef(metadata, body))
    }
  } else if (word.indexOf('ㅇ') !== -1) {
    const trailer = word.slice(1)
    if (trailer !== '') {
      // AS.ArgRef
      const relA = stack.pop()
      if (relA == null) throw Error('Internal::parseToken::EMPTY_STACK')
      const relF = +trailer
      stack.push(new AS.ArgRef(metadata, relA, relF))
    } else {
      // AS.FunRef
      const relF = stack.pop()
      if (relF == null) throw Error('Internal::parseToken::EMPTY_STACK')
      if (!(relF instanceof AS.Literal)) {
        throw SyntaxError(
          `함수 참조 시에는 정수 리터럴만 허용되는데 ${relF}를 받았습니다.`
        )
      }
      stack.push(new AS.FunRef(metadata, Number(relF.value)))
    }
  } else {
    stack.push(new AS.Literal(metadata, parseNumber(word)))
  }
}

/* Parses program into abstract syntax */
export function parse(filename: string, sentence: string) {
  const tokens = tokenize(filename, sentence)
  const stack: AS.AST[] = []
  for (const token of tokens) {
    parseToken(token, stack)
  }
  return stack
}
